Yafc.Model/Model/ProductionTable.cs

```diff
@@ -157,6 +162,32 @@ public void AddRecipe(RecipeOrTechnology recipe, IComparer<Goods> ingredientVari
             }
         }
 
+        private static EntityCrafter? GetSelectedFuelCrafter(RecipeOrTechnology recipe, Goods? selectedFuel) =>
+            selectedFuel?.fuelFor.OfType<EntityCrafter>()
+                .Where(e => e.recipes.Contains(recipe))
+                .AutoSelect(DataUtils.FavoriteCrafter);
+
+        private static EntityCrafter? GetSpentFuelCrafter(RecipeOrTechnology recipe, Goods? spentFuel) {
+            if (spentFuel is null) {
+                return null;
+            }
+            return recipe.crafters
+                .Where(c => c.energy.fuels.OfType<Item>().Any(e => e.fuelResult == spentFuel))
+                .AutoSelect(DataUtils.FavoriteCrafter);
+        }
+
+        private static Goods? GetFuelForSpentFuel(Goods? spentFuel, RecipeRow recipeRow) {
+            if (spentFuel is null) {
+                return null;
+            }
+            return recipeRow?.entity?.energy.fuels.Where(e => spentFuel?.miscSources.Contains(e) ?? false)
+                .AutoSelect(DataUtils.FavoriteFuel);
+        }
+
+        private static Goods? GetSelectedFuel(Goods? selectedFuel, RecipeRow recipeRow) =>
+            // Skipping AutoSelect since there will only be one result at most.
+            recipeRow?.entity?.energy.fuels.FirstOrDefault(e => e == selectedFuel);
```

Comment:
```suggestion
            recipeRow.entity?.energy.fuels.FirstOrDefault(e => e == selectedFuel);
```
In both of these, you can also remove the ? after entity if you make them local functions.

Reply:
I've removed one ? by adding a [NotNull] attribute, but i'm unable to get rid of the other, C# gives me a warning when i do.

![afbeelding](https://github.com/user-attachments/assets/a0fbe17c-e5f7-428e-a28e-05d74bfd3161)

Reply:
Oh nvm, i think you meant both instances of `recipeRow?`, not the `?` next to `entity` :) Those i've got fixed.